### ISSUE TITLE
Generating VS2015 project on Linux fails

### ISSUE THREAD

**Issue by afiefh** [mesonbuild/meson#3382]
Attempting to build a VS2015 project for my VS users fails on Linux:

        $ ~/tmp/bin/python3 ~/tmp/meson/meson/meson.py build/vs2015/ --backend vs2015

        Traceback (most recent call last):
        File "/home/afiefh/tmp/meson/meson/mesonbuild/mesonmain.py", line 369, in run
        app.generate()
        File "/home/afiefh/tmp/meson/meson/mesonbuild/mesonmain.py", line 151, in generate
        self._generate(env)
        File "/home/afiefh/tmp/meson/meson/mesonbuild/mesonmain.py", line 208, in _generate
        g.generate(intr)
        File "/home/afiefh/tmp/meson/meson/mesonbuild/backend/vs2010backend.py", line 161, in generate
        projlist = self.generate_projects()
        File "/home/afiefh/tmp/meson/meson/mesonbuild/backend/vs2010backend.py", line 312, in generate_projects
        self.gen_vcxproj(target, projfile, uuid)
        File "/home/afiefh/tmp/meson/meson/mesonbuild/backend/vs2010backend.py", line 937, in gen_vcxproj
        extra_link_args += compiler.get_link_whole_for(linkname)
        File "/home/afiefh/tmp/meson/meson/mesonbuild/compilers/compilers.py", line 1072, in get_link_whole_for
        return ['-Wl,--whole-archive'] + args + ['-Wl,--no-whole-archive']
        TypeError: can only concatenate list (not "str") to list

I noticed that the function get_link_whole_for has a different behavior for VCC than it does for GNUCompiler, adding the same safeguard to the gnu compiler solves the issue:

		 def get_link_whole_for(self, args):
	+        if not isinstance(args, list):
	+            args = [args]
			 return ['-Wl,--whole-archive'] + args + ['-Wl,--no-whole-archive']

Before you ask, yes our project runs on Linux, our build runs on Linux with GCC, but some of our developers prefer to use Windows and VS for development by mounting the code to a Windows drive but still everything is built exclusively on Linux and not Windows.

**Comment by jpakkane** [comment]
If you have a working patch, please submit a PR.

**Comment by nirbheek** [comment]
You successfully nerd-sniped me into checking how much work it would be to add this setup to our CI. Turns out it's [not a lot in the beginning](https://github.com/mesonbuild/meson/files/1900641/lolvsonlinux.txt), but there's a few more tests that fail because they assume you can't use the Visual Studio backend on Linux.

We could add this to the CI if you take that patch and complete it, if it amuses @jpakkane as much as it amused me.

**Comment by nirbheek** [comment]
> some of our developers prefer to use Windows and VS for development by mounting the code to a Windows drive

Curious: why do they need to use the solutions in that case? Just for the file listing?

**Comment by mikkelfj** [comment]
Can't speak for OP, but if you have a meson project and you want to ensure that the source is portable, you want test cases close to the target user scenario.

**Comment by afiefh** [comment]
I thought a couple of times if I actually want to write this out or not... The level of convolution is a bit embarrassing, so here goes:

TL;DR: Our building happens on Linux server, some people want a VS project along with their build so they can open the project in VS, they cannot be bothered to generate a local VS project on their Windows machine.

Background on the product I work on: The product we develop will only ever run on Linux, we only compile with GCC and even the OS we run on is pretty locked down and controlled, so in theory we are using Meson mostly as a ninja generator and less as a configuration/os detector. Our previous build system was a handwritten Makefile and people were satisfied with this.

However, years ago this product was developed and run on Windows machines, and some developers here (whose years of experience is more than years in my life I think) are very much against change, so they are still on their Windows machines, mounting the work directory from their Linux station and using Visual Studio to write the code. Installing Meson for them and explaining how to generate a project didn't work, they don't want to deal with meson (building is already hidden behind a buid.py file that just builds all targets). As a compromise it was decided to generate a Visual Studio project along with the ninja files when building so that they would be able to continue their work without change. Since building happens on Linux, that is where the ninja files are being generated as well.
